xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -174,6 +208,55 @@ __device__ void compute_final_mult(
   }
 }
 
+template <typename scalar_t, typename vec_t, int kBlockSizeK, int kBlockSizeQ>
+//__device__ __forceinline__ void apply_masking(
+__device__ void apply_masking(
+    scalar_t s_delta[kBlockSizeQ][kBlockSizeK],
+    at::PhiloxCudaState philox_args,
+    int64_t global_offset,
+    int64_t N,
+    scalar_t p,
+    int64_t col_offset) {
+  // strategy: initialize the rng so that each element in the attention
+  // matrix has its own subsequence, so that we can easily retrieve
+  // the element during backward
+
+  curandStatePhilox4_32_10_t state;
+  auto seeds = at::cuda::philox::unpack(philox_args);
+
+  // we will always sample 4 random floats at a time
+  // as it's more efficient
+  constexpr int kSampled = 4;
```

Comment:
same in triton, I'm guessing that this is HW dependent (big random word cut into pieces ?)

Reply:
It shouldn't be much more expensive, because curand saves unused randoms and yields them on future calls, without going through generation.

Reply:
@ngimel indeed, but in the current setup we have to reset the state of the philox generation very often (almost for every element), and this would make up for the slowdown, as we would reset the state (expensive), generate 4 floats and only use 1.

The strategy here is to reset the state only once every 4 elements in the output